### PR TITLE
Update sanitization filter to allow data: URIs in img tags

### PR DESCRIPTION
**Pull request by polybuildr**
Allowing data: URIs as values for the src attribute
adds great convenience without introducing any
security issues.

### CODE REVIEW

**Comment by polybuildr** [comment]
As per a discussion at https://github.com/github/markup/issues/270, this should be safe enough.

**Comment by kbrock** [comment]
nice :+1:

**Comment by jch** [comment]
@mastahyeti I see you were referenced in the other thread. Any thoughts on
the security implications of this change?

**Comment by mastahyeti** [comment]
There were a number of concerns brought up in https://github.com/github/markup/issues/270. I think we'd need to suss those out before making this change for GitHub.com. Specifically:

- We don't want links using the `data:` scheme and we add links to image tags by default.
- "What happens if a data: img goes out in an email? Do most clients handle it correctly?"
- "The only thing about base:64 that makes me worried, is the length of issues and logs, once it might get bigger(specially if it is a screenshot or something like this)."

**Comment by mhenry** [comment]
If supporting `data:*` is scary, is it possible to only support `data:image/png`, `data:image/gif`, and `data:image/gif`? Also, are there any mitigations for the concerns @mastahyeti copied over from github/markup#270? The spotty email client support for inline data URIs seems like the biggest doozy to me.

**Comment by traxair** [comment]
Any news on this and how soon it can be integrated? How do you enforce `data:image/png` instead of `data:*`? Only one line was modified in the PR, should be a piece of cake... right?

**Comment by zbeekman** [comment]
does this mean that I can't use base64 encoded svg images in README.md files?

**Comment by muroc** [comment]
I urge you, good and courageous people, to click "Merge" with all haste.

**Comment by corneadoug** [comment]
Would love that feature too, so that we don't have to host images

**Comment by JamieMcDonnell** [comment]
We are looking for support of this feature too - using the following markdown format is still not displaying the base64 image as it should:
`an example red dot`

Please get this live ASAP as it will assist development no end to get a visual when a user reports an issue with countless apps that rely on GIT.

Thanks all.

**Comment by tc0312** [comment]
@jch Could you kindly merge this PR or explain your concern? Thanks :)

**Comment by eugene-goroschenya** [comment]
+1

File: lib/html/pipeline/sanitization_filter.rb [hunk 1]
```diff
@@ -81,7 +81,7 @@ class SanitizationFilter < Filter
           'ins'        => {'cite' => ['http', 'https', :relative]},
           'q'          => {'cite' => ['http', 'https', :relative]},
           'img'        => {
-            'src'      => ['http', 'https', :relative],
+            'src'      => ['http', 'https', 'data', :relative],
```
Comment (astipkp) [inline lib/html/pipeline/sanitization_filter.rb]: (My data)

**Comment by JamieMcDonnell** [comment]
Hi all, any idea of an ETA when this will actually be live? Thanks in advance.

**Comment by CareyJWilliams** [comment]
Any further thoughts/discussion here?

**Comment by daetal-us** [comment]
👍

**Comment by jeanfw** [comment]
👍

**Comment by RichardBronosky** [comment]
I think it's time to reopen discussion on why this is not being accepted.

**Comment by CH-RhyMoore** [comment]
Me too.

**Comment by CH-RhyMoore** [comment]
Me too.

I was playing around with SVGs in Github Flavored Markdown today and ran into this limitation (again; I'd forgotten I'd run into several years ago). I suppose there's an additional security angle, but...I wish this worked. 

Here's my use case. I am working on a design system and there's a component for our set of icons. With something visual, it's often essential to have...a visual. But we're not committing the SVG files of our icons as individual assets. With our setup, this would be both unnecessary and noisy; individual files are created during builds and published, but not tracked as source code in the repo.

Wouldn't it be neat, I thought, if you could just see the icons in the README? No need for me to deploy a site. And if I could inline them, I wouldn't have to add noise to our diffs by committing assets just to display them, we wouldn't have to worry about files hosted elsewhere going poof, etc. This approach also happens to work very nicely with the tools I'm using to build and plans I've got for the future.

So, I tried URL encoding an SVG and using a data URI. The idea is that I could just write the icons directly to my README and, unsurprisingly, it does not work.

upload

```
![upload](data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2224%22%20height%3D%2224%22%20display%3D%22block%22%20pointer-events%3D%22none%22%20viewBox%3D%220%200%2024%2024%22%3E%3Cpath%20fill%3D%22none%22%20stroke%3D%22currentColor%22%20stroke-linejoin%3D%22round%22%20stroke-width%3D%221.5%22%20d%3D%22M22.25%2016v4.25c0%201.176-.895%202-2%202H3.75c-1.104%200-2-.824-2-2V16M12%202v16m5.75-10.527L12.001%201.75%206.25%207.473%22%2F%3E%3C%2Fsvg%3E)
```

For a little more info and examples of how this works outside of GitHub, see https://css-tricks.com/probably-dont-base64-svg.

**Comment by CH-RhyMoore** [comment]
Here's a demo with a script inside the SVG. https://codepen.io/morewry/pen/rdQevJ?editors=1001

I checked this in IE 10, IE 11, Yandex 14.12, Opera 52, Opera 51, Safari 8, Safari 9.1, Safari 10.1, Safari 11, Firefox 58, Firefox 59, Chrome 64, Chrome 65, Edge 15, and Edge 16.  No console output appears.

**Comment by SephReed** [comment]
Sure would be nice if it wasn't impossible to show all the icons of my icons repo in the README.  What's worse is that if this weren't a paid for private repo, it'd be a lot easier.

**Comment by akosthekiss** [comment]
I hope to give this PR a push with a ping. (The branch is now in a conflicting state but it's easy to resolve. Happy to help, if that helps. May it be the only thing that blocks this from merging.)

**Comment by polybuildr** [comment]
My impression was that this was blocked because of other reasons - not just because of the stale PR, but I've merged `master` now and resolved the conflict just in case.

Repository owners, please let me know if you'd like me to squash and rebase the commit - happy to do that too.

**Comment by vearutop** [comment]
Let the merge of this PR be a True Christmas Miracle!

**Comment by kivikakk** [comment]
I'm not sure who has write access and who doesn't (I don't), but as has been discussed in several other places: this gem is no longer the basis of HTML processing for github.com, and so changes made here will not be seen on GitHub itself.

**Comment by vearutop** [comment]
@kivikakk then why https://github.com/github/markup/issues/270 is closed as off-topic? What/where should be changed to allow embedding images in markdown media?

**Comment by kivikakk** [comment]
My last comment on the issue is exactly what explains why it's closed as off-topic, and tells you exactly where to reach out to. Thanks.

**Comment by miyconst** [comment]
It's mid of 2019 and the feature is not yet there.. quite disappointed, since embedding tiny Base64 images into markdown documents is a very convenient feature.

**Comment by kcak11** [comment]
Any updates regarding when this will get merged. Its been a long time and everyone seems to be waiting . . .

**Comment by SerialDev** [comment]
> I'm not sure who has write access and who doesn't (I don't), but as has been discussed in several other places: this gem is no longer the basis of HTML processing for github.com, and so changes made here will not be seen on GitHub itself.


I do have to say closing it as off topic and suggesting a reach-out to github.com/contact just completely removes any ability for people to track down whether this feature will eventually be implemented, and an issue tracker is also a great way to have some transparency. its been 5 years since this support has been requested, so I don't feel writing to ask what is going on?

**Pull request closed without merging**